Support for cascading deletes of part tables

zfj1 opened the pull request:
This PR addresses datajoint/datajoint-python#151, to allow cascading deletes that affect part tables to delete from the master, thus maintaining the many-to-one integrity. Also addresses #126 by preventing deletes from part tables directly.

A unit test is included under [tests/TestMasterPart.m](tests/TestMasterPart.m) with an example schema (adapted from datajoint/datajoint-python#934) under [tests/test_schemas/+Acquisition](tests/test_schemas/+Acquisition).

Also affected:
 - added `target` property for auto-populated tables to match the python version and improve usability of master/part tables (see [Acquisition.Segment](tests/test_schemas/+Acquisition/Segment.m))
 - fixed an apparent bug in dj.Relvar.del line 103 (#379)

guzman-raphael commented:
@zfj1 Thank you for your contribution! This PR looks great! We appreciate the effort! :clap:

**Quick development process note:** DataJoint MATLAB PR's are a bit tricky since they require a MATLAB license to run the tests with GitHub Actions. For now, this requires 2 PR's at a minimum: 1 into `stage` from your fork (remote PR) and 1 into `master` (local PR). Would you re-point your PR to `stage` so we can merge it to allow the tests to be run in a local PR? Once merged I can open a local PR for you from `stage->master` and we can conduct the proper code review there (will link it once ready).

Thank you for your patience with this process. We hope to improve the contribution process in the future but for now this is how it is done.

zfj1 commented:
@guzman-raphael Sorry for the delay, I've re-based to stage and made a few changes.

I ran into a bug while using this PR where part tables with matching dependent attributes in the secondary key would throw an error on deletion. To resolve this, I switched from using the datajoint native join (natural join) to an inner join on the master primary key (note: just for the deletion of part tables). This change has the added benefit of detecting renamed keys in the part tables if they exist.

dimitri-yatsenko commented on +dj/Schema.m:
```diff
@@ -43,9 +43,9 @@
-        tierPrefixes = {'#', '', '_', '__', '~'}
-        tierClasses = {'dj.Lookup' 'dj.Manual' 'dj.Imported' 'dj.Computed' 'dj.Jobs'}
+        allowedTiers = {'lookup' 'manual' 'imported' 'computed' 'job', 'hidden'}
+        tierPrefixes = {'#', '', '_', '__', '~', '$'}
```
`~` already denotes hidden tables such as Jobs, External, and Log (in python). We do not need to introduce a new table type/prefix.

dimitri-yatsenko commented on +dj/Hidden.m:
```diff
@@ -0,0 +1,12 @@
+classdef Hidden < dj.internal.UserRelation
```
Jobs and external tables should be of this type then.

dimitri-yatsenko commented on +dj/+internal/AutoPopulate.m:
```diff
@@ -52,7 +53,22 @@
     
     methods
-        
+        function target = getTarget(self)
```
can this be more naturally orrganized as a dependent property?

zfj1 replied:
Thanks for the suggestion. I've now implemented this as a getter for the `target_` property. Unfortunately, [Matlab's inheritance rules](https://www.mathworks.com/help/matlab/matlab_oop/modify-superclass-properties.html) don't allow for overrides so there's no simpler way I know of to implement this.

zfj1 commented:
@dimitri-yatsenko I've addressed your change requests. Some commits for another feature I'm working on implementing (https://github.com/datajoint/datajoint-python/issues/325) accidentally made their way into this branch, so I've removed those from this PR.
